pkg/util/pki/asn1_util.go

```diff
@@ -0,0 +1,190 @@
+	}
+
+	parts := strings.Split(oidString, ".")
+
+	oid = make(asn1.ObjectIdentifier, 0, len(parts))
+	for _, part := range parts {
+		value, err := strconv.Atoi(part)
+		if err != nil {
+			return nil, err
+		}
+
+		oid = append(oid, value)
+	}
+
+	return oid, nil
+}
+
+type UniversalValue struct {
+	Bytes           []byte
+	IA5String       string
+	UTF8String      string
+	PrintableString string
+}
+
+func MarshalUniversalValue(uv UniversalValue) ([]byte, error) {
+	// Make sure we have only one field set
+	{
+		var count int
+		if uv.Bytes != nil {
+			count++
+		}
+		if uv.IA5String != "" {
+			count++
+		}
+		if uv.UTF8String != "" {
+			count++
+		}
+		if uv.PrintableString != "" {
+			count++
+		}
+		if count != 1 {
+			return nil, fmt.Errorf("exactly one field must be set")
+		}
+	}
+
+	var bytes []byte
+
+	if uv.Bytes != nil {
+		bytes = uv.Bytes
+	} else {
+		rawValue := asn1.RawValue{
+			Class:      asn1.ClassUniversal,
+			IsCompound: false,
+		}
+		switch {
+		case uv.IA5String != "":
+			if err := isIA5String(uv.IA5String); err != nil {
+				return nil, errors.New("asn1: invalid IA5 string")
+			}
+			rawValue.Tag = asn1.TagIA5String
+			rawValue.Bytes = []byte(uv.IA5String)
+		case uv.UTF8String != "":
+			if !utf8.ValidString(uv.UTF8String) {
+				return nil, errors.New("asn1: invalid UTF-8 string")
+			}
+			rawValue.Tag = asn1.TagUTF8String
+			rawValue.Bytes = []byte(uv.UTF8String)
+		case uv.PrintableString != "":
+			if !isPrintable(uv.PrintableString) {
+				return nil, errors.New("asn1: invalid PrintableString string")
+			}
+			rawValue.Tag = asn1.TagPrintableString
+			rawValue.Bytes = []byte(uv.PrintableString)
+		}
+
+		universalBytes, err := asn1.Marshal(rawValue)
+		if err != nil {
+			return nil, err
+		}
+		bytes = universalBytes
+	}
+
+	return bytes, nil
+}
+
+func isIA5String(s string) error {
+	for _, r := range s {
+		// Per RFC5280 "IA5String is limited to the set of ASCII characters"
+		if r > unicode.MaxASCII {
+			return fmt.Errorf("x509: %q cannot be encoded as an IA5String", s)
+		}
+	}
+
+	return nil
+}
+
+// isPrintable reports whether the given b is in the ASN.1 PrintableString set.
+// '*' and '&' are also allowed, reflecting existing practice.
+func isPrintable(s string) bool {
+	for _, b := range s {
+		if 'a' <= b && b <= 'z' ||
+			'A' <= b && b <= 'Z' ||
+			'0' <= b && b <= '9' ||
+			'\'' <= b && b <= ')' ||
+			'+' <= b && b <= '/' ||
+			b == ' ' ||
+			b == ':' ||
+			b == '=' ||
+			b == '?' ||
+			// This is technically not allowed in a PrintableString.
+			// However, x509 certificates with wildcard strings don't
+			// always use the correct string type so we permit it.
+			b == '*' ||
+			// This is not technically allowed either. However, not
+			// only is it relatively common, but there are also a
+			// handful of CA certificates that contain it. At least
+			// one of which will not expire until 2027.
+			b == '&' {
+			continue
+		}
+
+		return false
+	}
+
+	return true
+}
```

Comment:
suggestion: I think this is taken from https://github.com/golang/go/blob/c95fe91d0715dc0a8d55ac80a80f383c3635548b/src/encoding/asn1/asn1.go#L430 right? If so, it'd be worth linking to the upstream function in a comment here.